cpp/src/arrow/compute/kernels/scalar_cast_string.cc

```diff
@@ -305,19 +310,198 @@ BinaryToBinaryCastExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* ou
       ctx, input, out->array_data().get());
 }
 
+// View -> Span
+template <typename O, typename I>
+enable_if_t<is_binary_view_like_type<I>::value && is_base_binary_type<O>::value, Status>
+BinaryToBinaryCastExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+  using OutputBuilderType = typename TypeTraits<O>::BuilderType;
+  const CastOptions& options = checked_cast<const CastState&>(*ctx->state()).options;
+  const ArraySpan& input = batch[0].array;
+
+  if constexpr (!I::is_utf8 && O::is_utf8) {
+    if (!options.allow_invalid_utf8) {
+      InitializeUTF8();
+      ArraySpanVisitor<I> visitor;
+      Utf8Validator validator;
+      RETURN_NOT_OK(visitor.Visit(input, &validator));
+    }
+  }
+
+  // XXX: a more efficient implementation that zero-copies the validity bitmap
+  // is possible, but requires a more complex implementation for building the
+  // offsets and data buffers
```

Comment:
Because of a possible input offset? Even without zero-copying, `CopyBitmap` would be much faster than appending validity bits one by one. That said, this can be left to a another GH issue (if we care at all).

Reply:
+1, please CopyBitmap the validity buffer

Reply:
How can I do that while still using the builders? I will create an issue and link it from the comments.

Reply:
`ArrayBuilder::UnsafeAppendToBitmap` is relied upon for essential things like updating the length.

https://github.com/apache/arrow/blob/66cb7495d1a43f3539cf66f6d88bac40fb9d28d4/cpp/src/arrow/array/builder_base.h#L227-L232

Reply:
https://github.com/apache/arrow/issues/43573

Reply:
> How can I do that while still using the builders? I will create an issue and link it from the comments.

You could instead use individual TypedBufferBuilders for the offsets and data buffers. That shouldn't add a lot of complexity.

Reply:
I'm not against it and I think this is necessary, but can it be done after this PR is merged? I mentioned this tip on the issue and pinged people that might be happy to help implementing this.

Reply:
Sounds fine to me, thanks.